### PR TITLE
[TS migration] Migrate 'SelectionScraper' lib to TypeScript

### PR DESCRIPTION
**Pull request by Skalakid**
### Details

Migrated SelectionScraper files to TypeScript
### Fixed Issues

$ https://github.com/Expensify/App/issues/24879
PROPOSAL: https://github.com/Expensify/App/issues/24879

### Tests

1. Log in
2. Enter any chat
3. Write few messages with various formatting
4. Copy all text inside the chat
5. Paste it to the main composer
6. Send message
7. Verify if all text has been copied and all formatings are correct
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as Text
### QA Steps

Same as Text
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/39538890/64d8e9d2-badc-4362-8150-0322f02b04be

</details>

<details>
<summary>Mobile Web - Chrome</summary>

N/A
</details>

<details>
<summary>Mobile Web - Safari</summary>

N/A
</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/39538890/f4b2f92d-bf9e-4d4d-984d-14afe64a50d4

</details>

<details>
<summary>iOS</summary>

N/A
</details>

<details>
<summary>Android</summary>

N/A
</details>

### CODE REVIEW

File: src/libs/SelectionScraper/index.ts [hunk 1]
```diff
@@ -64,17 +64,17 @@ const getHTMLOfSelection = () => {
             // and finally commonAncestorContainer.parentNode.closest('data-testid') is targeted dom.
             if (range.commonAncestorContainer instanceof HTMLElement) {
                 parent = range.commonAncestorContainer.closest(`[${tagAttribute}]`);
```
Comment (blazejkustra) [inline src/libs/SelectionScraper/index.ts:66]: Type parent, now it's of type any

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/libs/SelectionScraper/index.ts [hunk 2]
```diff
@@ -44,7 +45,7 @@ const getHTMLOfSelection = () => {
 
         // If clonedSelection has no text content this data has no meaning to us.
         if (clonedSelection.textContent) {
-            let parent;
+            let parent: globalThis.Element | null = null;
```
Comment (blazejkustra) [inline src/libs/SelectionScraper/index.ts:48]:
```suggestion
            let parent: Element | undefined;
```

File: src/libs/SelectionScraper/index.ts [hunk 3]
```diff
@@ -1,25 +1,26 @@
...
+import {DataNode, Element, Node} from 'domhandler';
 import CONST from '../../CONST';
+import GetCurrentSelection from './types';
 
 const elementsWillBeSkipped = ['html', 'body'];
 const tagAttribute = 'data-testid';
 
 /**
  * Reads html of selection. If browser doesn't support Selection API, returns empty string.
- * @returns {String} HTML of selection as String
```
Comment (blazejkustra) [inline src/libs/SelectionScraper/index.ts:12]:
```suggestion
 * @returns HTML of selection as String
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]

**Review by blazejkustra — APPROVED** [review]

**Review by parasharrajat — APPROVED** [review]

**Comment by danieldoglas** [comment]
@blazejkustra can you please do the checklist?

**Comment by blazejkustra** [comment]
@danieldoglas Because this PR is from SWM, Callstack engineers will make the checklist for us. cc @fabioh8010

File: src/libs/SelectionScraper/index.ts [hunk 4]
```diff
@@ -1,25 +1,27 @@
...
+ * @returns HTML of selection as String
  */
-const getHTMLOfSelection = () => {
+const getHTMLOfSelection = (): string => {
     // If browser doesn't support Selection API, return an empty string.
-    if (!window.getSelection) {
+    const selection = window.getSelection();
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/SelectionScraper/index.ts:17]: @Skalakid is OOO this week, he will adjust it once he's back

File: src/libs/SelectionScraper/index.ts [hunk 5]
```diff
@@ -64,17 +66,17 @@ const getHTMLOfSelection = () => {
             // and finally commonAncestorContainer.parentNode.closest('data-testid') is targeted dom.
             if (range.commonAncestorContainer instanceof HTMLElement) {
                 parent = range.commonAncestorContainer.closest(`[${tagAttribute}]`);
-            } else {
-                parent = range.commonAncestorContainer.parentNode.closest(`[${tagAttribute}]`);
+            } else if (range.commonAncestorContainer.parentNode) {
+                parent = (range.commonAncestorContainer.parentNode as HTMLElement).closest(`[${tagAttribute}]`);
             }
```
Comment (fabioh8010) [inline src/libs/SelectionScraper/index.ts:71]:
```suggestion
            } else {
                parent = (range.commonAncestorContainer.parentNode as HTMLElement | null)?.closest(`[${tagAttribute}]`) ?? null;
            }
```

Would be safer if we do like this, in this way we are not changing conditional flow.

File: src/libs/SelectionScraper/index.ts [hunk 6]
```diff
@@ -64,17 +66,17 @@ const getHTMLOfSelection = () => {
...
 
             // Keep traversing up to clone all parents with 'data-testid' attribute.
             while (parent) {
                 const cloned = parent.cloneNode();
                 cloned.appendChild(child);
-                child = cloned;
+                child = cloned as DocumentFragment;
 
-                parent = parent.parentNode.closest(`[${tagAttribute}]`);
+                parent = (parent.parentNode as HTMLElement).closest(`[${tagAttribute}]`);
```
Comment (fabioh8010) [inline src/libs/SelectionScraper/index.ts:79]:
```suggestion
                parent = (parent.parentNode as HTMLElement | null)?.closest(`[${tagAttribute}]`) ?? null;
```

`parentNode` can be null.

File: src/libs/SelectionScraper/index.ts [hunk 7]
```diff
@@ -96,58 +98,61 @@ const getHTMLOfSelection = () => {
...
     }
 
     return {
         ...dom,
-        data,
         name: domName,
         attribs: domAttribs,
         children: domChildren,
-    };
+    } as Element;
```
Comment (fabioh8010) [inline src/libs/SelectionScraper/index.ts:147]: I didn't fully understand why this change was made.

**Comment by blazejkustra** [comment]
@Skalakid is OOO this week, he will respond & adjust the code once he is back

File: src/libs/SelectionScraper/index.ts (hunk 7 above)
Comment (Skalakid) [reply src/libs/SelectionScraper/index.ts:147]: I made this change because `replaceNodes` function should return `Node` type elements. `Node` type is a general type and there are many more specific ones that extend it. The problem was that children can change different types of Nodes. According to the `domhandler` types there isn't any type that have both `data` and `attribs`/`children` attributes. So if there is Node have `data` I'm returning `DataNode`, if no I'm returning `Element` type Node

File: src/libs/SelectionScraper/index.ts (hunk 6 above)
Comment (Skalakid) [reply src/libs/SelectionScraper/index.ts:79]: Good catch, thanks!

File: src/libs/SelectionScraper/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Skalakid) [reply src/libs/SelectionScraper/index.ts:17]: According to the Window types, `getSelection` should be always present. So I think it won't throw

File: src/libs/SelectionScraper/index.ts [hunk 8]
```diff
@@ -96,58 +100,59 @@ const getHTMLOfSelection = () => {
...
-    let domName = dom.name;
-    let domChildren;
-    const domAttribs = {};
-    let data;
-
+const replaceNodes = (dom: Node, isChildOfEditorElement: boolean): Node => {
     // Encoding HTML chars '< >' in the text, because any HTML will be removed in stripHTML method.
-    if (dom.type === 'text') {
-        data = Str.htmlEncode(dom.data);
+    const domDataNode = dom as DataNode;
```
Comment (blazejkustra) [inline src/libs/SelectionScraper/index.ts:106]: @Skalakid Why this logic was moved up?

Comment (blazejkustra) [reply src/libs/SelectionScraper/index.ts:106]: I wouldn't change logic as it looks quite fragile 😅

**Review by blazejkustra — APPROVED** [review]

File: src/libs/SelectionScraper/index.ts [hunk 9]
```diff
@@ -138,16 +143,15 @@ const replaceNodes = (dom, isChildOfEditorElement) => {
         name: domName,
         attribs: domAttribs,
         children: domChildren,
-    };
+    } as unknown as Node;
```
Comment (blazejkustra) [inline src/libs/SelectionScraper/index.ts:146]: @fabioh8010 This assertion was my idea, other thing that we could do is to return `new Element(...)` or `new DataNode(...)` instead of assertion, but I didn't want to change the logic that much, wdyt?

Comment (fabioh8010) [reply src/libs/SelectionScraper/index.ts:146]:
```suggestion
    } as Element & DataNode;
```

How about this? :)

Comment (blazejkustra) [reply src/libs/SelectionScraper/index.ts:146]: Definitely better

**Review by fabioh8010 — APPROVED** [review]

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/libs/SelectionScraper/index.ts [hunk 10]
```diff
@@ -96,40 +100,41 @@ const getHTMLOfSelection = () => {
...
- * @param {Boolean} isChildOfEditorElement
- * @returns {Object} htmlparser2 dom representation
+ * @param dom - dom htmlparser2 dom representation
  */
-const replaceNodes = (dom, isChildOfEditorElement) => {
-    let domName = dom.name;
-    let domChildren;
-    const domAttribs = {};
-    let data;
+const replaceNodes = (dom: Node, isChildOfEditorElement: boolean): Node => {
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/SelectionScraper/index.ts:105]: `Node` is a parent class of `DataNode` and `Element`, this function has to be more generic that's why `Node` is the return type

File: src/libs/SelectionScraper/types.ts [hunk 11]
```diff
@@ -0,0 +1,3 @@
+type GetCurrentSelection = () => string;
+
+export default GetCurrentSelection;
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/SelectionScraper/types.ts:3]: Usually in other libs`types.ts` files we used default export, so for consistency I would leave it as is @parasharrajat

File: src/libs/SelectionScraper/index.ts [hunk 12]
```diff
@@ -96,40 +100,41 @@ const getHTMLOfSelection = () => {
...
+    if (dom.type.toString() === 'text' && dom instanceof DataNode) {
         data = Str.htmlEncode(dom.data);
     }
 
-    // We are skipping elements which has html and body in data-testid, since ExpensiMark can't parse it. Also this data
-    // has no meaning for us.
-    if (dom.attribs && dom.attribs[tagAttribute]) {
-        if (!elementsWillBeSkipped.includes(dom.attribs[tagAttribute])) {
-            domName = dom.attribs[tagAttribute];
+    if (dom instanceof Element) {
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/SelectionScraper/index.ts:116]: This condition is added to [narrow down the type](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) so we can use some of the properties in the if body.

> This will allow the function to return garbage value

Not exactly sure what you mean by this? We tested the logic and the argument is either `DataNode` or `Element`, so depending on the argument type the code from first/second if is executed

File: src/libs/SelectionScraper/index.ts [hunk 13]
```diff
@@ -96,40 +100,41 @@ const getHTMLOfSelection = () => {
...
-    let data;
+const replaceNodes = (dom: Node, isChildOfEditorElement: boolean): Node => {
+    let domName;
+    let domChildren: Node[] = [];
+    const domAttribs: Element['attribs'] = {};
+    let data = '';
 
     // Encoding HTML chars '< >' in the text, because any HTML will be removed in stripHTML method.
-    if (dom.type === 'text') {
+    if (dom.type.toString() === 'text' && dom instanceof DataNode) {
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/SelectionScraper/index.ts:112]: > Previously we were strictly matching type property. Now we are doing a type conversion before matching it. I don't know if it changes anything logically but why do we have to change this condition?

@Skalakid Use this instead:
```ts
import { ElementType } from "domelementtype";
...

if (dom.type === ElementType.Text && dom instanceof DataNode) {
```

> We defined dom as Node on the function definition and now we are checking the type of DataNode. Either we should define the correct type or not.

`DataNode` is a child of `Node`, so it's perfectly fine to narrow down the type like we do in the if statement `dom instanceof DataNode` @parasharrajat, wdyt?

`dom instanceof DataNode` is needed to narrow down the type & use it's additional properties that aren't available on `Node`

File: src/libs/SelectionScraper/index.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/SelectionScraper/index.ts:105]: `name` is only available on `Element`

File: src/libs/SelectionScraper/index.ts [hunk 14]
```diff
@@ -96,40 +100,41 @@ const getHTMLOfSelection = () => {
 
 /**
  * Clears all attributes from dom elements
- * @param {Object} dom htmlparser2 dom representation
- * @param {Boolean} isChildOfEditorElement
- * @returns {Object} htmlparser2 dom representation
+ * @param dom - dom htmlparser2 dom representation
  */
-const replaceNodes = (dom, isChildOfEditorElement) => {
-    let domName = dom.name;
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/SelectionScraper/index.ts:104]: name is only available on Element

File: src/libs/SelectionScraper/index.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/SelectionScraper/index.ts:116]: We need to narrow the type, but you point is also valid. I'm proposing something like this: 
```ts
    if (dom.type === ElementType.Text && dom instanceof DataNode) {
        ...
    } else if (dom instanceof Element) {
        ...
    } else {
        throw new Error(`Unknown dom type: ${dom.type}`);
    }
```

We would have to test it thoroughly, but it's the best what can be done. Any other ideas? @parasharrajat @Skalakid

[the remark this answers is not in the document]
Comment (Skalakid) [reply src/libs/SelectionScraper/index.ts]: I agree with that

**Comment by Skalakid** [comment]
Sorry, I missed the notification about this PR, thanks @blazejkustra for answering the comments. I've aded proposed changes

**Review by parasharrajat — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by Skalakid** [comment]
@danieldoglas bump :D

**Comment by danieldoglas** [comment]
Hey guys, I'm back from OOO now.  We're in a conference at this moment, so I'll hold for 3 more days to merge this since I don't want to risk generating any regressions in the middle of a conference. Sorry for the long delay here!

**Review by danieldoglas — DISMISSED** [review]

**Comment by danieldoglas** [comment]
oops, lint is failing after merge. Let's merge this today, sorry for the delay!

**Review by danieldoglas — APPROVED** [review]
